[DOCS-9644] Standalone Error Tracking for Mobile (#26569)
DataDog/documentation: content/en/real_user_monitoring/mobile_and_tv_monitoring/setup/android.md

---
## Overview

> The Android SDK can be used to instrument your application for both [Real User Monitoring (RUM)][1] and [Error Tracking][2].

Reviewer: ```suggestion
You can use the Android SDK to instrument your application for both [Real User Monitoring (RUM)][1] and [Error Tracking][2].
```

The Datadog Android SDK supports Android 5.0+ (API level 21) and Android TV.

## Setup

1. Declare Datadog SDK as a dependency.
2. Specify application details in the UI.
3. Initialize Datadog SDK with application context.
4. Enable the feature to start sending data.
5. Initialize the Interceptor to track network events.

### Declare the Datadog SDK as a dependency

> Declare [dd-sdk-android-rum][3] and the [Gradle plugin][4] as a dependency in your **application module's** `build.gradle` file.

Reviewer: ```suggestion
Declare [dd-sdk-android-rum][3] and the [Gradle plugin][4] as dependencies in your **application module's** `build.gradle` file.
```

```

### Specify application details in the UI
{{< tabs >}}
{{% tab "RUM" %}}

1. Navigate to [**Digital Experience** > **Add an Application**][1].
2. Select `android` as the application type and enter an application name to generate a unique Datadog application ID and client token.
3. To instrument your web views, click the **Instrument your webviews** toggle. For more information, see [Web View Tracking][2].

> 4. To disable automatic user data collection for either client IP or geolocation data, use the toggle button for those settings. For more information, see [RUM Android Data Collected][3].

Reviewer: ```suggestion
4. To disable automatic user data collection for either client IP or geolocation data, use the toggles for those settings. For more information, see [RUM Android Data Collected][3].
```

   {{< img src="real_user_monitoring/android/android-new-application.png" alt="Create a RUM application for Android in Datadog" style="width:90%;">}}

[1]: https://app.datadoghq.com/rum/application/create
[2]: /real_user_monitoring/android/web_view_tracking/
[3]: /real_user_monitoring/android/data_collected/

{{% /tab %}}
{{% tab "Error Tracking" %}}

1. Navigate to [**Error Tracking** > **Settings** > **Browser and Mobile** > **Add an Application**][1].
2. Select `android` as the application type and enter an application name to generate a unique Datadog application ID and client token.
3. To instrument your web views, click the **Instrument your webviews** toggle. For more information, see [Web View Tracking][2].

> 4. To disable automatic user data collection for either client IP or geolocation data, use the toggle button for those settings. For more information, see [Android Data Collected][3].

Reviewer: ```suggestion
4. To disable automatic user data collection for either client IP or geolocation data, use the toggles for those settings. For more information, see [Android Data Collected][3].
```

   {{< img src="real_user_monitoring/error_tracking/mobile-new-application.png" alt="Create a RUM application for Android in Datadog" style="width:90%;">}}

[1]: https://app.datadoghq.com/error-tracking/settings/setup/client
[2]: /real_user_monitoring/android/web_view_tracking/
[3]: /real_user_monitoring/android/data_collected/

{{% /tab %}}
{{< /tabs >}}

> To ensure the safety of your data, you must use a client token. If you used only [Datadog API keys][5] to configure the Datadog SDK, they would be exposed client-side in the Android application's APK byte code. 

Reviewer: ```suggestion
To ensure the safety of your data, you must use a client token. If you use only [Datadog API keys][5] to configure the Datadog SDK, they are exposed client-side in the Android application's APK byte code. 
```

For more information about setting up a client token, see the [Client Token documentation][6].

### Initialize Datadog SDK with application context

In the initialization snippet, set an environment name, service name, and version number. In the examples below, `APP_VARIANT_NAME` specifies the variant of the application that generates data. For more information, see [Using Tags][7].

> See [`trackingConsent`][8] to add GDPR compliance for your EU users, and [other configuration options][9] to initialize the library.

Reviewer: ```suggestion
See [`trackingConsent`][8] to add GDPR compliance for your EU users. See [other configuration options][9] to initialize the library.
```

1. Navigate to [**Digital Experience** > **Add an Application**][1].
2. Select `android` as the application type and enter an application name to generate a unique Datadog application ID and client token.
3. To instrument your web views, click the **Instrument your webviews** toggle. For more information, see [Web View Tracking][2].

> 4. To disable automatic user data collection for either client IP or geolocation data, use the toggles for those settings. For more information, see [RUM Android Data Collected][3].

Reviewer: ```suggestion
4. To disable automatic user data collection for client IP or geolocation data, use the toggles for those settings. For more information, see [RUM Android Data Collected][3].
```

1. Navigate to [**Error Tracking** > **Settings** > **Browser and Mobile** > **Add an Application**][1].
2. Select `android` as the application type and enter an application name to generate a unique Datadog application ID and client token.
3. To instrument your web views, click the **Instrument your webviews** toggle. For more information, see [Web View Tracking][2].

> 4. To disable automatic user data collection for either client IP or geolocation data, use the toggles for those settings. For more information, see [Android Data Collected][3].

Reviewer: ```suggestion
4. To disable automatic user data collection for client IP or geolocation data, use the toggles for those settings. For more information, see [Android Data Collected][3].
```

{{< /tabs >}}
{{< /site-region >}}

> The initialization credentials require your application's variant name and uses the value of `BuildConfig.FLAVOR`. With the variant, the SDK can match the errors reported from your application with the mapping files uploaded by the Gradle plugin. If you do not have variants, the credentials use an empty string. 

Reviewer: ```suggestion
The initialization credentials require your application's variant name and use the value of `BuildConfig.FLAVOR`. With the variant, the SDK can match the errors reported from your application to the mapping files uploaded by the Gradle plugin. If you do not have variants, the credentials use an empty string. 
```

The Gradle plugin automatically uploads the appropriate ProGuard `mapping.txt` file at build time so you can view deobfuscated error stack traces. For more information, see the [Track Android Errors][10].

### Sample sessions

> To control the data your application sends to Datadog, you can specify a sample rate for sessions while [initializing the RUM or Error Tracking][11] as a percentage between 0 and 100.

Reviewer: ```suggestion
To control the data your application sends to Datadog, you can specify a sample rate for sessions when [initializing RUM or Error Tracking][11]. The sample rate is a percentage between 0 and 100.
```

- `TrackingConsent.PENDING`: (Default) The SDK starts collecting and batching the data but does not send it to the
 collection endpoint. The SDK waits for the new tracking consent value to decide what to do with the batched data.
- `TrackingConsent.GRANTED`: The SDK starts collecting the data and sends it to the data collection endpoint.
- `TrackingConsent.NOT_GRANTED`: The SDK does not collect any data. You are not able to manually send any logs, traces, or events.

To update the tracking consent after the SDK is initialized, call `Datadog.setTrackingConsent(<NEW CONSENT>)`. The SDK changes its behavior according to the new consent. For example, if the current tracking consent is `TrackingConsent.PENDING` and you update it to:

- `TrackingConsent.GRANTED`: The SDK sends all current batched data and future data directly to the data collection endpoint.
- `TrackingConsent.NOT_GRANTED`: The SDK wipes all batched data and does not collect any future data.

> ### Enable feature to start sending data

Reviewer: ```suggestion
### Enable the feature to start sending data
```

{{< tabs >}}
{{% tab "Kotlin" %}}

```kotlin
   val rumConfig = RumConfiguration.Builder(applicationId)
     .trackInteractions()
     .trackLongTasks(durationThreshold) // Not applicable to Error Tracking
     .useViewTrackingStrategy(strategy)
     .build()
   Rum.enable(rumConfig)
```

{{% /tab %}}
{{% tab "Java" %}}

```java
   RumConfiguration rumConfig = new RumConfiguration.Builder(applicationId)
     .trackInteractions()
     .trackLongTasks(durationThreshold)  // Not applicable to Error Tracking
     .useViewTrackingStrategy(strategy)
     .build();
   Rum.enable(rumConfig);
```

{{% /tab %}}
{{< /tabs >}}

See [`ViewTrackingStrategy`][12] to enable automatic tracking of all your views (activities, fragments, and more).

### Initialize the Interceptor to track network events

> 1. If you want to have distributed tracing, add and enable Trace feature, see [Datadog Android Trace Collection documentation][13] to learn how.

Reviewer: ```suggestion
1. If you want to have distributed tracing, [add and enable the Trace feature][13].
```

{{% /tab %}}
{{< /tabs >}}

> This records each request processed by the `OkHttpClient` as a resource, with all the relevant information automatically filled (URL, method, status code, and error). Only the network requests that started when a view is active are tracked. To track requests when your application is in the background, [create a view manually][15].

Reviewer: ```suggestion
This records each request processed by the `OkHttpClient` as a resource, with all the relevant information (URL, method, status code, and error) automatically filled in. Only the network requests that started when a view is active are tracked. To track requests when your application is in the background, [create a view manually][15].
```

### `Closeable` extension

> You can monitor `Closeable` instance usage by using `useMonitored` method, which reports any error happened to Datadog and close the resource afterwards.

Reviewer: ```suggestion
You can monitor `Closeable` instance usage with the `useMonitored` method, which reports errors to Datadog and closes the resource afterwards.
```

## Sending data when device is offline

The Android SDK ensures availability of data when your user device is offline. In case of low-network areas, or when the device battery is too low, all events are first stored on the local device in batches. 

Each batch follows the intake specification. Batches are sent as soon as the network is available, and the battery is high enough to ensure the Datadog SDK does not impact the end user's experience. If the network is not available while your application is in the foreground, or if an upload of data fails, the batch is kept until it can be sent successfully.
 
This means that even if users open your application while offline, no data is lost. To ensure the SDK does not use too much disk space, the data on the disk is automatically discarded if it gets too old.

## Further Reading

{{< partial name="whats-next/whats-next.html" >}}

> [1]:/real_user_monitoring/

Author: ```suggestion
[1]: /real_user_monitoring/
```
